### ISSUE TITLE
[java] Possible new rule for non-reifiable varargs methods?

### ISSUE THREAD

**Issue by Fernal73** [pmd/pmd#2918]
**Description:**

Can we have a rule that flags non-reifiable `varargs` static and non-constructor method declarations that don't have the `@SafeVarargs` annotation or `@SuppressWarnings({"unchecked", "varargs"})` annotation? 

https://docs.oracle.com/javase/8/docs/technotes/guides/language/non-reifiable-varargs.html#suppressing

**Comment by oowekyala** [comment]
> Property to enforce one of @SafeVarargs, @SuppressWarnings or both.

`@SafeVarargs` is superior to SuppressWarnings as warnings are not propagated to all use sites. I think we should enforce only this, with an explanation

**Comment by Fernal73** [comment]
Yup, missed that. Modified above!

**Comment by oowekyala** [comment]
We should also note that `@SafeVarargs` cannot be applied to non-final non-static methods, eg interface methods. I think in that case, the possible fixes we should recommend are
* use a list as parameter (and you can use `List.of(...)`)
* make the method static
* push the method down to an abstract class and make it final

Examples from our codebase:

https://github.com/pmd/pmd/blob/dfb5c4e2d270cfa3e4524a6cd78113b886d5caac/pmd-core/src/main/java/net/sourceforge/pmd/lang/ast/Node.java#L265-L269

https://github.com/pmd/pmd/blob/51ca1a5961868dfe59549579d9eb862ceec09bdc/pmd-core/src/main/java/net/sourceforge/pmd/lang/ast/NodeStream.java#L1087


IMO this rule would be a better fit for the Design category

**Comment by Fernal73** [comment]
> IMO this rule would be a better fit for the Design category

Perfect.

**Comment by Fernal73** [comment]
```java
public static <T> T[] unsafe(T... elements) {
    return elements; 
// unsafe! don't ever return a parameterized varargs array
}
```
https://www.baeldung.com/java-safevarargs

We could also consider adding a check for the above to the rule since it deals with using `varargs` safely.

**Comment by Fernal73** [comment]
```java
// The WRONG way to use varargs to pass one or more arguments!
static int min(int... args) {
if (args.length == 0)
throw new IllegalArgumentException("Too few arguments");
int min = args[0];
for (int i = 1; i < args.length; i++)
if (args[i] < min)
min = args[i];
return min;
}
```

Can we also implement a check for the above on a best efforts basis?
 
Example from Joshua Bloch's Effective Java.

**Comment by oowekyala** [comment]
> We could also consider adding a check for the above to the rule since it deals with using varargs safely.

I think this should be a related rule but not the same. The implementation would be clearly disjoint, the documentation too. If you care about this, I think it would deserve its own issue

> Can we also implement a check for the above on a best efforts basis?

Would be nice, but for the same reason, not in the same rule

**Comment by Fernal73** [comment]
> We should also note that `@SafeVarargs` cannot be applied to non-final non-static methods, eg interface methods. I think in that case, the possible fixes we should recommend are
> 
> * use a list as parameter (and you can use `List.of(...)`)
> * make the method static
> * push the method down to an abstract class and make it final
https://stackoverflow.com/questions/37829160/safevarargs-on-interface-method/3795902
